instrumentation-genai/opentelemetry-instrumentation-google-genai/src/opentelemetry/instrumentation/google_genai/tool_call_wrapper.py

```diff
@@ -0,0 +1,190 @@
+# Copyright The OpenTelemetry Authors
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import functools
+import inspect
+from typing import Any, Callable, Optional, Union
+
+from google.genai.types import (
+    ToolListUnion,
+    ToolListUnionDict,
+    ToolOrDict,
+)
+
+from opentelemetry import trace
+from opentelemetry.semconv._incubating.attributes import (
+    code_attributes,
+)
+
+from .flags import is_content_recording_enabled
+from .otel_wrapper import OTelWrapper
+
+ToolFunction = Callable[..., Any]
+
+
+def _to_otel_value(python_value):
+    """Coerces parameters to something representable with Open Telemetry."""
+    if python_value is None:
+        return None
+    if isinstance(python_value, list):
+        return [_to_otel_value(x) for x in python_value]
+    if isinstance(python_value, dict):
+        return {
+            key: _to_otel_value(val) for (key, val) in python_value.items()
+        }
+    if hasattr(python_value, "model_dump"):
+        return python_value.model_dump()
+    if hasattr(python_value, "__dict__"):
+        return _to_otel_value(python_value.__dict__)
+    return repr(python_value)
+
+
+def _create_function_span_name(wrapped_function):
+    """Constructs the span name for a given local function tool call."""
+    function_name = wrapped_function.__name__
+    return f"tool_call {function_name}"
+
+
+def _create_function_span_attributes(
+    wrapped_function, function_args, function_kwargs, extra_span_attributes
+):
+    """Creates the attributes for a tool call function span."""
+    result = {}
+    if extra_span_attributes:
+        result.update(extra_span_attributes)
+    result[code_attributes.CODE_FUNCTION_NAME] = wrapped_function.__name__
+    result["code.module"] = wrapped_function.__module__
+    result["code.args.positional.count"] = len(function_args)
+    result["code.args.keyword.count"] = len(function_kwargs)
+    return result
+
+
+def _record_function_call_argument(
+    span, param_name, param_value, include_values
+):
+    attribute_prefix = f"code.function.parameters.{param_name}"
+    type_attribute = f"{attribute_prefix}.type"
+    span.set_attribute(type_attribute, type(param_value).__name__)
+    if include_values:
+        value_attribute = f"{attribute_prefix}.value"
+        span.set_attribute(value_attribute, _to_otel_value(param_value))
```

Comment:
It looks like `_to_otel_value()` can return a dict or heterogeneous array, which will get dropped (until we support complex attributes) https://opentelemetry.io/docs/specs/otel/common/#attribute. I'm not sure the best course of action, maybe jsonifying

Reply:
Done. I JSON-ified where it wasn't otherwise acceptable as an attribute value.